File: haskell/makeBoard.hs
module MakeBoard
(
    getBoard
    randomList
) where

import System.Random (randomRIO)

randomList :: Int -> Int -> Int -> IO([Int])
randomList s e 0 = return []
randomList s e n = do
  r  <- randomRIO (s,e)
  rs <- randomList s e (n-1)
  if not (r `elem` rs) then return (r:rs)
  else randomList s e n

pattern :: Int -> Int -> Int
pattern r c = (3*(r`mod`3)+r `quot` 3+c)`mod`9

sumIn :: Int -> [Int] -> [Int]
sumIn 0 l = l
sumIn n [] = []
sumIn n l = (n*3)+(head l) : (sumIn n (tail l))

build :: [Int] -> [Int] -> [Int]
build [] ys = []
build xs [] = xs
build (x:xs) ys = (sumIn x ys) ++ (build xs ys)

makeBoard :: [Int] -> [Int] -> [Int] -> [[Int]]
makeBoard [] cs ns = []
makeBoard rs [] ns = [rs]
makeBoard (r:rs) (c:cs) ns = [getPattern r (c:cs) ns] ++ makeBoard rs (c:cs) ns

getPattern :: Int -> [Int] -> [Int] -> [Int]
getPattern r [] ns = []
getPattern r (c:cs) ns = (ns!!(pattern r c)) : getPattern r cs ns

getNum :: Int -> Int -> [Int] -> Int
getNum r c ns = ns!!(pattern r c)

getSample = do
    l <- randomList 0 2 3
    return l

getNums = do
    l <- randomList 1 9 9
    return l

getBoard = do
    list1 <- getSample
    list2 <- getSample
    let rows = (build list1 list2)

    list1 <- getSample
    list2 <- getSample
    let columns = (build list1 list2)

    nums <- getNums
    let board = makeBoard rows columns nums
    return board

File: prolog/makeBoard.pl
base(3).
side(9).
rbase([0,1,2]).
nums(N) :- random_permutation([1,2,3,4,5,6,7,8,9], N).

pattern(R, C, P) :-
  base(B),
  side(S),
  P is (B*(R mod B)+R//B+C) mod S.

rows(R) :-
  rbase(B),
  random_permutation(B, J),
  random_permutation(B, K),
  build(J, K, R).

cols(C) :-
  rbase(B),
  random_permutation(B, J),
  random_permutation(B, K),
  build(J, K, C).

sumIn(_, [] , []).
sumIn([], _, []).
sumIn([J|JT], [K|KT] , L) :-
  base(B),
  sumIn([J|JT], KT, L1),
  N is J*B+K,
  L = [N | L1].

build(_, [] , []).
build([], _, []).
build([J|JT], [K|KT] , L) :-
  sumIn([J|JT], [K|KT] , L1),
  build(JT, [K|KT], L2),
  append(L1, L2, L).

getBoard(_, [], _, []).
getBoard([], _, _, []).
getBoard([R|RT], [C|CT], N, B):-
  pattern(R, C, P),
  getBoard([R|RT], CT, N, B1),
  nth0(P, N, E),
  B = [E | B1].

makeBoard(_, [], _, []).
makeBoard([], _, _, []).
makeBoard([R|RT], [C|CT], N, B):-
  getBoard([R|RT], [C|CT], N, B1),
  makeBoard(RT, [C|CT], N, B2),
  append([B1], B2, B).

board(B):-
  nums(N),
  rows(R),
  cols(C),
  makeBoard(R, C, N, B).
